python/ray/data/_internal/block_batching/iter_batches.py

```diff
@@ -0,0 +1,118 @@
+The algorithm uses both pipeline parallelism and data parallelism:
+
+If prefetch_batches=2, these are all the batches in flight:
+
+[User thread] trains on Batch 0
+   - [Fetch thread] Batch 1 in output queue
+        - [Worker thread 1] Batch 2 formatting + collating
+        - [Worker thread 2] Batch 3 formatting + collating
+        - [Raylet] Batches 4 + 5 fetched to local object store memory
+
+At any point in time there are prefetch_batches+1 batches in local heap memory.
+And the next set of prefetch_batches in local object store memory.
+
+The actual steps are as follows:
+
+In a single async thread, do the following:
+    1. Trigger Ray local prefetching of `prefetch_batches` worth of block object
+        references.
+    2. Resolve (i.e. call `ray.get()`) on the block references.
+    3. Perform the necessary batch slicing to construct full batches, possibly
+        shuffling if necessary.
+    4. Then, in a threadpool consisting of `prefetch_batches` threads:
+        3. Format the batches to the provided batch format.
+        4. Apply the collate function.
+    5. Fetch outputs from the threadpool, maintaining order of the batches.
+"""
+
+
+def prefetch_batches_locally(
+    block_ref_iter: Iterator[Tuple[ObjectRef[Block], BlockMetadata]],
+    prefetcher: BlockPrefetcher,
+    num_batches_to_prefetch: int,
+    batch_size: Optional[int],
+) -> Iterator[ObjectRef[Block]]:
+    """Given an iterator of batched block references, returns an iterator over the same
+    block references while prefetching `num_batches_to_prefetch` batches in advance.
+
+    Args:
+        block_ref_iter: An iterator over batched block references.
+        prefetcher: The prefetcher to use.
+        num_batches_to_prefetch: The number of batches to prefetch ahead of the
+            current batch during the scan.
+        batch_size: User specified batch size, or None to let the system pick.
+    """
+
+    sliding_window = collections.deque()
+    current_window_size = 0
+
+    if batch_size:
+        num_rows_to_prefetch = num_batches_to_prefetch * batch_size
+    else:
+        num_rows_to_prefetch = None
+
+    # Create and fetch the initial window.
+    while True:
+        try:
+            next_block_ref_and_metadata = next(block_ref_iter)
+            sliding_window.append(next_block_ref_and_metadata)
+            current_window_size += next_block_ref_and_metadata[1].num_rows
+        except StopIteration:
+            break
+        # Stop adding if the number of rows in this window is greater than
+        # requested batch size.
+        if batch_size and current_window_size >= num_rows_to_prefetch:
+            break
+        # Stop adding if batch_size is None and the number of blocks in this window
+        # is greater than requested batches to prefetch.
+        elif not batch_size and len(sliding_window) >= num_batches_to_prefetch:
+            break
+
+    prefetcher.prefetch_blocks([block_ref for block_ref, _ in list(sliding_window)])
```

Comment:
Don't we need some stats instrumentation here to have parity? E.g.
```suggestion
    with stats.iter_wait_s.timer() if stats else nullcontext():
        prefetcher.prefetch_blocks([block_ref for block_ref, _ in list(sliding_window)])
```

Reply:
I don't think `ray.wait` time is very meaningful for users and am planning to remove it in the new codepath...

Prefetching is a non-blocking operation

Reply:
Ah cool, I forgot that reworking the status is in-scope so parity isn't a focus. That sounds good to me!

Reply:
Yep! Old codepath is still using it for parity